extensions/wrapper/clients/typescript-client/src/oauth2/OAuthMiddleware.ts

```diff
@@ -0,0 +1,34 @@
+import {Middleware, RequestContext, ResponseContext} from '../generated';
+import {needsAuthentication} from '../utils/HttpUtils';
+import {AccessTokenService} from './AccessTokenService';
+import {injectAccessTokenHeader} from './RequestUtils';
+
+export class OAuthMiddleware {
```

Comment:
Why can't we use functions here? A middleware is only created once. You could do state via variable scopes:

````ts
export function buildOauth2Middleware(clientCredentials: ClientCredentials): Middleware {
   const accessTokenService = new AccessTokenService(clientCredentials);
   return {
      pre: async (ctx) => {
         const accessToken = await accessTokenService.getAccessToken();
         // ...
      },
      post: async (ctx) => { 
         // ...
      }
   }
}

````